spark-cassandra-connector/src/test/scala/org/apache/spark/sql/cassandra/PredicatePushDownSpec.scala

```diff
@@ -0,0 +1,214 @@
+package org.apache.spark.sql.cassandra
+
+import org.scalatest.{Matchers, FlatSpec}
+
+import com.datastax.spark.connector.cql._
+import com.datastax.spark.connector.types.IntType
+
+class PredicatePushDownSpec extends FlatSpec with Matchers {
+
+  // We don't want this test to rely on any Spark code,
+  // so we're using our own Filters
+  trait Filter
+  case class EqFilter(columnName: String) extends Filter
+  case class InFilter(columnName: String) extends Filter
+  case class RangeFilter(columnName: String) extends Filter
+  case object UnsupportedFilter extends Filter
+
+  implicit object FilterOps extends PredicateOps[Filter] {
+    override def columnName(p: Filter) = p match {
+      case EqFilter(name) => name
+      case InFilter(name) => name
+      case RangeFilter(name) => name
+      case UnsupportedFilter => throw new IllegalArgumentException("Unsupported predicate")
+    }
+    override def isRangePredicate(p: Filter) = p.isInstanceOf[RangeFilter]
+    override def isSingleColumnPredicate(p: Filter) = p != UnsupportedFilter
+    override def isEqualToPredicate(p: Filter) = p.isInstanceOf[EqFilter]
+    override def isInPredicate(p: Filter) = p.isInstanceOf[InFilter]
+  }
+
+  val pk1 = ColumnDef("pk1", PartitionKeyColumn, IntType)
+  val pk2 = ColumnDef("pk2", PartitionKeyColumn, IntType)
+  val c1 = ColumnDef("c1", ClusteringColumn(0), IntType)
+  val c2 = ColumnDef("c2", ClusteringColumn(1), IntType)
+  val c3 = ColumnDef("c3", ClusteringColumn(2), IntType)
+  val i1 = ColumnDef("i1", RegularColumn, IntType, indexed = true)
+  val i2 = ColumnDef("i2", RegularColumn, IntType, indexed = true)
+  val r1 = ColumnDef("r1", RegularColumn, IntType)
+  val r2 = ColumnDef("r2", RegularColumn, IntType)
+
+  val table = TableDef(
+    keyspaceName = "test",
+    tableName = "test",
+    partitionKey = Seq(pk1, pk2),
+    clusteringColumns = Seq(c1, c2, c3),
+    regularColumns = Seq(i1, i2, r1, r2)
+  )
+
+  "PredicatePushDown" should "push down an equality partition key predicate" in {
+    val f1 = EqFilter("pk1")
+    val f2 = EqFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1, f2), table)
+    ppd.predicatesToPushDown should contain allOf(f1, f2)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "not push down a partition key predicate for a part of the partition key (1)" in {
+    val f1 = EqFilter("pk1")
+    val ppd = new PredicatePushDown(Set[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain only f1
+  }
+
+  it should "not push down a partition key predicate for a part of the partition key (2)" in {
+    val f1 = EqFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain only f1
+  }
+
+  it should "not push down a range partition key predicate" in {
```

Comment:
the previous two tests are not needed, because this tests test the same paths, just with different input. If the tested path is the same, the number of test doing that does not bring more confidence.